pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/InMemoryDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,193 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.delayed;
+
+import io.netty.util.Timeout;
+import io.netty.util.Timer;
+import io.netty.util.TimerTask;
+
+import java.time.Clock;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.bookkeeper.mledger.impl.PositionImpl;
+import org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers;
+import org.apache.pulsar.common.util.collections.TripleLongPriorityQueue;
+
+@Slf4j
+public class InMemoryDelayedDeliveryTracker implements DelayedDeliveryTracker, TimerTask {
+
+    private final TripleLongPriorityQueue priorityQueue = new TripleLongPriorityQueue();
+
+    private final PersistentDispatcherMultipleConsumers dispatcher;
+
+    // Reference to the shared (per-broker) timer for delayed delivery
+    private final Timer timer;
+
+    // Current timeout or null if not set
+    private Timeout timeout;
+
+    // Timestamp at which the timeout is currently set
+    private long currentTimeoutTarget;
+
+    private final long tickTimeMillis;
+
+    private final Clock clock;
+
+    InMemoryDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispatcher, Timer timer, long tickTimeMillis) {
+        this(dispatcher, timer, tickTimeMillis, Clock.systemUTC());
+    }
+
+    InMemoryDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispatcher, Timer timer, long tickTimeMillis, Clock clock) {
+        this.dispatcher = dispatcher;
+        this.timer = timer;
+        this.tickTimeMillis = tickTimeMillis;
+        this.clock = clock;
+    }
+
+    @Override
+    public boolean addMessage(long ledgerId, long entryId, long deliveryAt) {
+        long now = clock.millis();
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Add message {}:{} -- Delivery in {} ms ", dispatcher.getName(), ledgerId, entryId,
+                    deliveryAt - now);
+        }
+        if (deliveryAt < now) {
+            // It's already about time to deliver this message
+            return false;
+        }
+
+        priorityQueue.add(deliveryAt, ledgerId, entryId);
+        updateTimer();
+        return true;
+    }
+
+    /**
+     * Return true if there's at least a message that is scheduled to be delivered already
+     */
+    @Override
+    public boolean hasMessageAvailable() {
+        return !priorityQueue.isEmpty() && priorityQueue.peekN1() <= clock.millis();
+    }
+
+    /**
+     * Get a set of position of messages that have already reached
+     */
+    @Override
+    public Set<PositionImpl> getScheduledMessages(int maxMessages) {
+        int n = maxMessages;
+        Set<PositionImpl> positions = new TreeSet<>();
```

Comment:
if we already know life cycle of `PositionImpl` then can we use `PositionImplRecylce` instead?

Reply:
I wanted to keep it simple for now. We can iteratively improve and optimize.